src/DataStreams/PostgreSQLBlockInputStream.cpp

```diff
@@ -0,0 +1,286 @@
+#include "PostgreSQLBlockInputStream.h"
+
+#if USE_LIBPQXX
+#include <Columns/ColumnNullable.h>
+#include <Columns/ColumnString.h>
+#include <Columns/ColumnArray.h>
+#include <Columns/ColumnsNumber.h>
+#include <Columns/ColumnDecimal.h>
+#include <DataTypes/IDataType.h>
+#include <DataTypes/DataTypeNullable.h>
+#include <DataTypes/DataTypeArray.h>
+#include <DataTypes/DataTypesDecimal.h>
+#include <Interpreters/convertFieldToType.h>
+#include <IO/ReadHelpers.h>
+#include <IO/WriteHelpers.h>
+#include <IO/ReadBufferFromString.h>
+#include <Common/assert_cast.h>
+#include <ext/range.h>
+#include <common/logger_useful.h>
+
+namespace DB
+{
+
+namespace ErrorCodes
+{
+    extern const int BAD_ARGUMENTS;
+    extern const int UNKNOWN_TYPE;
+}
+
+
+PostgreSQLBlockInputStream::PostgreSQLBlockInputStream(
+    ConnectionPtr connection_,
+    const std::string & query_str_,
+    const Block & sample_block,
+    const UInt64 max_block_size_)
+    : query_str(query_str_)
+    , max_block_size(max_block_size_)
+    , connection(connection_)
+{
+    description.init(sample_block);
+    for (const auto idx : ext::range(0, description.sample_block.columns()))
+        if (description.types[idx].first == ValueType::vtArray)
+            prepareArrayInfo(idx, description.sample_block.getByPosition(idx).type);
+    /// pqxx::stream_from uses COPY command, will get error if ';' is present
+    if (query_str.ends_with(';'))
+        query_str.resize(query_str.size() - 1);
+}
+
+
+void PostgreSQLBlockInputStream::readPrefix()
+{
+    tx = std::make_unique<pqxx::read_transaction>(*connection);
+    stream = std::make_unique<pqxx::stream_from>(*tx, pqxx::from_query, std::string_view(query_str));
+}
+
+
+Block PostgreSQLBlockInputStream::readImpl()
+{
+    /// Check if pqxx::stream_from is finished
+    if (!stream || !(*stream))
+        return Block();
+
+    MutableColumns columns = description.sample_block.cloneEmptyColumns();
+    size_t num_rows = 0;
+
+    while (true)
+    {
+        const std::vector<pqxx::zview> * row{stream->read_row()};
+
+        if (!row)
+        {
+            /// row is nullptr if pqxx::stream_from is finished
+            stream->complete();
```

Comment:
I assume if destructor will be called between our InputStream processing it will be safe http://pqxx.org/development/libpqxx/. 
Would it be better to put this code for consistency in readSuffix ?

Reply:
Probably yes, good point, thank you.